apps/astarte_pairing/lib/astarte_pairing/queries.ex

```diff
@@ -338,4 +332,111 @@ defmodule Astarte.Pairing.Queries do
         {:error, :database_connection_error}
     end
   end
+
+  @spec custom_query(Xandra.statement(), String.t() | nil, Xandra.values(), [query_opt()]) ::
+          {:ok, term()} | {:error, term()}
+  def custom_query(statement, realm \\ nil, params \\ %{}, opts \\ []) do
+    {custom_opts, query_opts} = parse_opts(opts)
+
+    Xandra.Cluster.run(@cluster, fn conn ->
+      execute_query(conn, statement, realm, params, query_opts, custom_opts)
+    end)
+  end
+
+  defp execute_query(conn, statement, realm, params, query_opts, custom_opts) do
+    with {:ok, prepared} <- prepare_query(conn, statement, realm) do
+      case Xandra.execute(conn, prepared, params, query_opts) do
+        {:ok, page} ->
+          cast_query_result(page, custom_opts)
+
+        {:error, error} ->
+          %{message: message, tag: tag} = database_error_message(error, custom_opts[:context])
+
+          _ = Logger.warn(message, tag: tag)
+
+          {:error, :database_error}
+      end
+    end
+  end
+
+  defp use_realm(_conn, nil = _realm), do: :ok
+
+  defp use_realm(conn, realm) when is_binary(realm) do
+    with true <- Realm.valid_name?(realm),
+         {:ok, %Xandra.SetKeyspace{}} <- Xandra.execute(conn, "USE #{realm}") do
+      :ok
+    else
+      _ -> {:error, :realm_not_found}
+    end
+  end
+
+  defp prepare_query(conn, statement, realm) do
+    with :ok <- use_realm(conn, realm) do
+      case Xandra.prepare(conn, statement) do
+        {:ok, page} ->
+          {:ok, page}
+
+        {:error, reason} ->
+          _ = Logger.warn("Cannot prepare query: #{inspect(reason)}.", tag: "db_error")
+          {:error, :database_error}
+      end
+    end
+  end
+
+  defp parse_opts(opts) do
+    {custom_opts, query_opts} = Keyword.split(opts, Keyword.keys(@default_custom_query_opts))
+    query_opts = Keyword.merge(@default_query_opts, query_opts)
+    custom_opts = Keyword.validate!(custom_opts, @default_custom_query_opts)
+
+    {custom_opts, query_opts}
+  end
+
+  defp cast_query_result(page, opts) do
+    result_with_defaults =
+      case opts[:result] do
+        :first -> {:first, nil}
+        :first! -> {:first!, :not_found}
+        x -> x
+      end
+
+    case result_with_defaults do
+      :page ->
+        {:ok, page}
+
+      :list ->
+        {:ok, Enum.to_list(page)}
+
+      {:first, default} ->
+        {:ok, Enum.at(page, 0, default)}
+
+      {:first!, error} ->
+        Enum.at(page, 0)
+        |> case do
+          nil -> {:error, error}
+          first -> {:ok, first}
+        end
+    end
+  end
+
+  defp database_error_message(%Xandra.Error{message: message, reason: reason}, nil = _context) do
+    %{message: "Database error #{reason}: #{message}", tag: "db_error"}
```

Comment:
Since Xandra errors implement the Exception behaviour, you could use `Exception.message/1` rather than manually getting into the struct and retrieving message and reason